potassco/reify.h

```diff
@@ -0,0 +1,144 @@
+    Reifier& operator=(const Reifier&) = delete;
+
+    //! Parses an aspif program and writes facts to the output stream.
+    void parse(std::istream& in);
+
+    //! Writes the program header (e.g., incremental tag) as facts.
+    void initProgram(bool incremental) override;
+    //! Prepares for a new program step (currently does nothing).
+    void beginStep() override;
+    //! Writes a rule as facts
+    void rule(HeadType ht, AtomSpan head, LitSpan body) override;
+    //! Writes a rule as facts
+    void rule(HeadType ht, AtomSpan head, Weight_t bound, WeightLitSpan body) override;
+    //! Writes a minimize directive as facts.
+    void minimize(Weight_t prio, WeightLitSpan lits) override;
+    //! Writes a projection directive as facts.
+    void project(AtomSpan atoms) override;
+    //! Writes an output atom directive as facts.
+    void outputAtom(Atom_t atom, std::string_view name) override;
+    //! Writes an output term directive as facts.
+    void outputTerm(Id_t termId, std::string_view name) override;
+    //! Writes an output condition as facts.
+    void output(Id_t termId, LitSpan condition) override;
+    //! Writes an external directive as facts.
+    void external(Atom_t a, TruthValue v) override;
+    //! Writes assumption directive as facts.
+    void assume(LitSpan lits) override;
+    //! Writes a heuristic directive as facts.
+    void heuristic(Atom_t a, DomModifier t, int bias, unsigned prio, LitSpan condition) override;
+    //! Writes an edge directive as facts.
+    void acycEdge(int s, int t, LitSpan condition) override;
+    //! Writes a theory number term as facts.
+    void theoryTerm(Id_t termId, int number) override;
+    //! Writes a theory symbolic term as facts.
+    void theoryTerm(Id_t termId, std::string_view name) override;
+    //! Writes a theory compound term as facts.
+    void theoryTerm(Id_t termId, int cId, IdSpan args) override;
+    //! Writes a theory element as facts.
+    void theoryElement(Id_t elementId, IdSpan terms, LitSpan cond) override;
+    //! Writes a theory atom as facts.
+    void theoryAtom(Id_t atomOrZero, Id_t termId, IdSpan elements) override;
+    //! Writes a theory atom with guard as facts.
+    void theoryAtom(Id_t atomOrZero, Id_t termId, IdSpan elements, Id_t op, Id_t rhs) override;
+
+    //! At the end of each step, writes SCC facts and clears stepData_.
+    void endStep() override;
+
+private:
+    //! Compute SCCs for a given head and body literals.
+    template <typename L>
+    void calculateSccs(AtomSpan head, std::span<const L> body);
+    //! Print a fact to the output stream.
+    template <typename... T>
+    void printFact(const char* name, const T&... args);
+    //! Print a fact for the current step if step reification is enabled.
+    template <typename... T>
+    void printStepFact(const char* name, const T&... args);
+    //! Insert a tuple into a map and print a fact if it was not already present.
+    template <typename M, typename T>
```

Comment:
Please consider simplifying this interface and dropping the header-dependency to `std::vector`.
I.e. you can replace the overload sets with two functions taking a `std::span`:
```cpp
//! Insert a tuple into a map and print a fact if it was not already present.
template <typename M, typename T>
auto tuple(M& map, const char* name, std::span<T> args) -> size_t;
//! Insert a ordered tuple into a map and print a fact if it was not already present.
template <typename M, typename T>
auto orderedTuple(M& map, const char* name, std::span<T> args) -> size_t;
```
You can then remove the `#include <vector>` from the header.